build: Disable c++20 designator syntax in clang builds

Split from: https://github.com/envoyproxy/envoy/pull/11316#issuecomment-642675562

C99 style struct designator syntax is not part of the c++17 standard.  We should consider flipping compiler flags to disable its usage and prevent breakage under compilers other than clang.

Some combination of the following flags should do the trick:
-Wc99-designator
-Wc99-extensions
-Wc++20-designator (may require clang 10)
-Wc++20-extensions (may require clang 10)

Filing since I think this issue came up again in PR 11773

cc: @wrowe @sunjayBhatia

Enabling -Wc++20-extensions on clang builds detects the issue in #11316

We do use some C++17 extensions in the code, so enabling that version of the warning wouldn't work:

In file included from source/common/http/path_utility.cc:1:
In file included from bazel-out/k8-fastbuild/bin/source/common/http/_virtual_includes/path_utility_lib/common/http/path_utility.h:3:
bazel-out/k8-fastbuild/bin/include/envoy/http/_virtual_includes/header_map_interface/envoy/http/header_map.h:34:7: error: use of the 'fallthrough' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
      FALLTHRU;
      ^
bazel-out/k8-fastbuild/bin/source/common/common/_virtual_includes/macros/common/common/macros.h:50:20: note: expanded from macro 'FALLTHRU'
#define FALLTHRU [[fallthrough]]

--- a/bazel/envoy_internal.bzl
+++ b/bazel/envoy_internal.bzl
@@ -48,7 +48,7 @@ def envoy_copts(repository, test = False):
                repository + "//bazel:windows_fastbuild_build": [],
                repository + "//bazel:windows_dbg_build": [],
            }) + select({
-               repository + "//bazel:clang_build": ["-fno-limit-debug-info", "-Wgnu-conditional-omitted-operand"],
+               repository + "//bazel:clang_build": ["-fno-limit-debug-info", "-Wgnu-conditional-omitted-operand", "-Wc++20-extensions"],
                repository + "//bazel:gcc_build": ["-Wno-maybe-uninitialized"],
                "//conditions:default": [],
            }) + select({

